### PR TITLE
Add a SourceName field into RasterSources

### LINKED ISSUE #197: RasterSource Identification
Currently `RasterSource` is expected to be identified by `uri`:

Previously `uri`:
https://github.com/geotrellis/geotrellis-contrib/blob/b954cbe21cd770ea0b36a2230f2c11074ec664ab/vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala#L45

currently `dataPath`:
https://github.com/geotrellis/geotrellis-contrib/blob/aca902c5c9ff8e37169b6e8f95bc22decf7ae5e2/vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala#L48

However there are at least two cases where that is not possible:
- `InMemoryRasterSource`: https://github.com/geotrellis/geotrellis-contrib/pull/138/files
- `MosaicRasterSource`: https://github.com/geotrellis/geotrellis-contrib/blob/master/vlm/src/main/scala/geotrellis/contrib/vlm/MosaicRasterSource.scala

What we need a is a trait hierarchy to be mixed in with  implementations of `RasterSource` similar to what is discussed here: https://github.com/geotrellis/geotrellis-contrib/pull/138

This would allow the users of the interface to match against the mix-ins and pull out raster addresses if they're available.

### PR DESCRIPTION
**Pull request by pomadchin**
# Overview

This PR allows `RasterSources` to have a name (or identifier) different from the path, for the in memory cases or for the `MosaicRasterSource` that can't have a path to the source.

## Checklist

- [x] Add entry to CHANGELOG.md

Closes #197

### CODE REVIEW

File: vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala [hunk 1]
```diff
@@ -42,9 +42,7 @@ import java.util.ServiceLoader
   * @groupdesc reproject Functions to resample raster data in target projection.
   * @groupprio reproject 2
   */
-trait RasterSource extends CellGrid[Long] with RasterMetadata with Serializable {
-  def dataPath: DataPath
-
+trait RasterSource extends CellGrid[Long] with NamedRasterSource with RasterMetadata with Serializable {
```
Comment (pomadchin) [inline vlm/src/main/scala/geotrellis/contrib/vlm/RasterSource.scala]: Should it be mixed in in this place? or it is better to create a separate `NamedRasterSource extends RasterSource`?

File: vlm/src/main/scala/geotrellis/contrib/vlm/effect/geotiff/BaseGeoTiffRasterSource.scala [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+import geotrellis.raster.CellType
+import geotrellis.raster.io.geotiff.reader.GeoTiffReader
+import geotrellis.raster.io.geotiff.{MultibandGeoTiff, Tags}
+import geotrellis.util.RangeReader
+
+import cats.Monad
+import cats.syntax.functor._
+import cats.syntax.apply._
+
+abstract class BaseGeoTiffRasterSource[F[_]: Monad: UnsafeLift] extends RasterSourceF[F] {
```
Comment (echeipesh) [inline vlm/src/main/scala/geotrellis/contrib/vlm/effect/geotiff/BaseGeoTiffRasterSource.scala]: I think we should avoid `Base_` raster sources. It doesn't really win much and adds a layer of indirection. I think in this case just duplicating those method implementation is perfectly reasonable.

File: gdal/src/main/scala/geotrellis/contrib/vlm/gdal/GDALPath.scala [hunk 3]
```diff
@@ -39,18 +39,18 @@ import java.net.MalformedURLException
  *
  *  @example "zip+s3://bucket/prefix/zipped-data.zip!data.tif"
  */
-case class GDALDataPath(path: String) extends DataPath
+case class GDALPath(value: String) extends SourcePath
```
Comment (echeipesh) [inline gdal/src/main/scala/geotrellis/contrib/vlm/gdal/GDALPath.scala]: 💯

File: vlm/src/main/scala/geotrellis/contrib/vlm/SourcePath.scala [hunk 4]
```diff
@@ -21,16 +21,17 @@ import java.net.URI
 /**
  * Represents the path to data that is to be read.
  */
-trait DataPath {
+trait SourcePath extends SourceName {
```
Comment (echeipesh) [inline vlm/src/main/scala/geotrellis/contrib/vlm/SourcePath.scala]: Yes, this makes a lot of sense.

**Comment by pomadchin** [comment]
Merging once CI is happy!

**Pull request merged**